### PR TITLE
Chainer: Simplify slstm test

### PR DESCRIPTION
**Pull request by dido1998**

### CODE REVIEW

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 1]
```diff
@@ -169,6 +170,87 @@ def test_double_backward(self,  backend_config):
...
             backend_config)
+"""
+
+
+@testing.parameterize(*testing.product_dict(
+    [
+        {'batch_size': 10, 'n_units': 4},
+        {'batch_size': 8, 'n_units': 5},
+        {'batch_size': 6, 'n_units': 24},
+        {'batch_size': 4, 'n_units': 15},
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: These parameterizations seem quite similar but are they testing different scenarios/code paths? I'd understand testing edge cases where values include e.g. 0.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 2]
```diff
@@ -169,6 +170,87 @@ def test_double_backward(self,  backend_config):
...
+    def setUp(self):
+        dtype = self.dtype
+
+        if dtype == numpy.float16:
+            self.check_forward_options.update({
+                'rtol': 1e-2, 'atol': 1e-2})
+            self.check_backward_options.update({
+                'rtol': 1e-2, 'atol': 1e-2})
+            self.check_double_backward_options.update(
+                {'rtol': 1e-2, 'atol': 1e-2})
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: Are the tolerance relaxations needed and if so would you mind commenting on thy?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 3]
```diff
@@ -169,6 +170,87 @@ def test_double_backward(self,  backend_config):
...
+
+    def forward(self, inputs, device):
+        c1, c2, x1, x2 = inputs
+        out = functions.slstm(c1, c2, x1, x2)
+        return out
+
+    def forward_expected(self, inputs):
+        c1, c2, x1, x2 = inputs
+        with chainer.using_config('use_ideep', 'never'):
+            out = functions.slstm(c1, c2, x1, x2)
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: It seems like you're computing the expected values using the functions which this test should be targeting. How about reusing the old code?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 4]
```diff
@@ -169,6 +170,87 @@ def test_double_backward(self,  backend_config):
         self.check_double_backward(
             self.inputs, self.grad_outputs, self.grad_grad_inputs,
             backend_config)
+"""
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: Could you remove the previous code instead of commenting it out?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 5]
```diff
@@ -169,6 +170,87 @@ def test_double_backward(self,  backend_config):
...
+        testing.product({
+            'use_cudnn': ['never'],
+        })
+        # With cuDNN
+        + testing.product({
+            'use_cudnn': ['always'],
+            'cudnn_deterministic': [True, False],
+            'autotune': [True, False],
+        })]))
+class TestSLSTM(testing.FunctionTestCase):
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: You can override `generate_grad_outputs` to test the previous variations in how `grad_outputs` are given.

Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: Could you somehow also keep the previous "flat" tests, e.g. using parameterization?

**Comment by hvy** [comment]
Thanks for the PR. Please take a look at my comments above. In general, it seems like the test coverage/variations are reduced quite a bit but it be good to keep it, at least on the same level as before unless there's a good reason not to do so.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py (hunk 1 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: I have added tests for different number of dimensions. Let me know if we can include more edge cases.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py (hunk 2 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: I am using the same tolerances that were used in the slstm test that is currently there in the master branch.

**Comment by dido1998** [comment]
I have made the changes. Let me know if any other changes are required.

**Comment by hvy** [comment]
Thanks for the quick fixed. What do you say about this one though https://github.com/chainer/chainer/pull/7805/files#r307117429 ?

**Comment by hvy** [comment]
Can you also fix the Travis failures?

**Comment by dido1998** [comment]
I saw that the flat tests convert the arrays to 2d so I added separate tests for 2d arrays. Is that okay?

**Comment by hvy** [comment]
Isn't it easier to add a `flat` parameter to the test class and condition based on it in `generate_inputs` and `generate_grad_outputs` ?

**Comment by dido1998** [comment]
I have added a flat parameter.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 6]
```diff
@@ -29,146 +29,126 @@ def inject_backend_tests(method_names):
...
-
-    def test_double_backward(self,  backend_config):
-        self.check_double_backward(
-            self.inputs, self.grad_outputs, self.grad_grad_inputs,
-            backend_config)
+    def generate_grad_outputs(self, outputs_template):
+        grad_out = []
+
+        h = outputs_template[0]
+        c = outputs_template[0]
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: The return values are `c` followed by `h` so it should be `h = outputs_template[1]` I think.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 7]
```diff
@@ -29,146 +29,126 @@ def inject_backend_tests(method_names):
...
+        x_shape.append(4 * self.shape[1])
+        for i in range(2, len(self.shape)):
+            x_shape.append(self.shape[i])
+
+        x_shape = tuple(x_shape)
+        c1 = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
+        c2 = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
+        x1 = numpy.random.uniform(-1, 1, x_shape).astype(self.dtype)
+        x2 = numpy.random.uniform(-1, 1, x_shape).astype(self.dtype)
+        print(self.flat)
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: Ops, debug print.
```suggestion
```

**Comment by dido1998** [comment]
I have removed the print and corrected the indexes.

**Comment by hvy** [comment]
You've mixed up the indices of `c` and `h`.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
You should be able to generate `None` gradients without overriding private methods and also not have to skip the double backward test after https://github.com/chainer/chainer/pull/7831.

**Comment by hvy** [comment]
#7831 has been merged. Could you update this PR to not override the private method?

**Comment by hvy** [comment]
Jenkins, test this please.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py [hunk 8]
```diff
@@ -29,146 +29,121 @@ def inject_backend_tests(method_names):
...
         if self.dtype == numpy.float16:
             self.check_forward_options = {'atol': 1e-3, 'rtol': 1e-2}
-            self.check_backward_options = {
-                'dtype': numpy.float64, 'atol': 5e-3, 'rtol': 5e-2}
-            self.check_double_backward_options = {
-                'dtype': numpy.float64, 'atol': 5e-3, 'rtol': 5e-2}
+            self.check_backward_options = {'atol': 5e-3, 'rtol': 5e-2}
+            self.check_double_backward_options = {'atol': 5e-3, 'rtol': 5e-2}
+        if self.grad_outputs[0] is False or self.grad_outputs[1] is False:
+            self.skip_double_backward_test = True
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: You don't need to skip the double backward test anymore. Could you remove these two lines above?

Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: Done.

Comment (hvy) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: Hm, it seems like the skip is still needed for old style functions (`Function` and not `FunctionNode`). Could you add the skip back, and maybe leave a TODO comment to remove the skip afterwards?

Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: I have made the changes.

Comment (hvy) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py]: You only need to skip double backward.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.
